### PR TITLE
[GEOS-5188] LegendDecoration does not respect target size

### PR DESCRIPTION
**Pull request by taba90**
<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Committs changing the REST API, or any configuration object, should check it the REST API docs (Swagger YAML files and classic documentation) need to be updated.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by taba90** [comment]
JIRA ticket https://osgeo-org.atlassian.net/browse/GEOS-5188

File: src/wms/src/main/java/org/geoserver/wms/decoration/LegendDecoration.java [hunk 1]
```diff
@@ -185,8 +122,8 @@ public Dimension findOptimalSize(Graphics2D g2d, WMSMapContent mapContext) {
 
     public void paint(Graphics2D g2d, Rectangle paintArea, WMSMapContent mapContext)
             throws Exception {
-        List<LayerLegend> legends = this.legends.get();
-
+        List<LayerLegend> legends =
+                this.legends.get() != null ? this.legends.get() : getLayerLegend(g2d, mapContext);
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/decoration/LegendDecoration.java]: Ok, so you need the legends to be computed, but they are normally done in the method above... and are making sure they are present even if the auto-sizing did not happen. Can you add a code comment to clarify it?

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetMapIntegrationTest.java [hunk 2]
```diff
@@ -1988,4 +1988,27 @@ public void testLayoutLegendStyleTextFitBox() throws Exception {
         sInfo.setLegend(null);
         catalog.save(sInfo);
     }
+
+    @Test
+    public void testLayoutLegendWithSpecifiedTargetSize() throws Exception {
+        File layouts = getDataDirectory().findOrCreateDir("layouts");
+        URL layout = GetMapIntegrationTest.class.getResource("../test-layout-with-size.xml");
+        FileUtils.copyURLToFile(layout, new File(layouts, "test-layout-with-size.xml"));
+        BufferedImage image =
```
Comment (aaime) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetMapIntegrationTest.java]: So here you're asking for a map that's actually smaller than the legend decoration, so you expect the decoration to cover all of it, map not even being visible, and the decoration be cut as it does not fit.
Please clarify it in a comment.

**Comment by taba90** [comment]
thanks for your review @aaime feedbacks have been applied

**Comment by aaime** [comment]
@taba90 merged, backport PRs welcomed

**Pull request merged**